angular/projects/global/src/lib/components/table-template/table-template-utils.ts

```diff
@@ -0,0 +1,87 @@
+import { Injectable } from '@angular/core';
+import { Router, Params } from '@angular/router';
+import { PlatformLocation } from '@angular/common';
+import {
+  TableObject,
+  DEFAULT_TABLE_PAGE_SIZE,
+  DEFAULT_TABLE_CURRENT_PAGE,
+  DEFAULT_TABLE_SORT_BY,
+  DEFAULT_TABLE_KEYWORDS
+} from './table-object';
+
+@Injectable()
+export class TableTemplateUtils {
+  constructor(private platformLocation: PlatformLocation, private router: Router) {}
+
+  public updateUrl(tableObject: TableObject) {
+    let currentUrl = this.router.url;
+    currentUrl = (this.platformLocation as any).getBaseHrefFromDOM() + currentUrl.slice(1);
+    currentUrl = currentUrl.split(';')[0];
+    currentUrl += `;currentPage=${tableObject.currentPage};pageSize=${tableObject.pageSize}`;
+    if (tableObject.keywords !== '') {
+      currentUrl += `;keywords=${tableObject.keywords}`;
+    }
+    if (tableObject.sortBy !== '' && tableObject.sortBy !== null) {
+      currentUrl += `;sortBy=${tableObject.sortBy}`;
+    }
+    if (tableObject.filter !== null && tableObject.filter !== {}) {
+      Object.keys(tableObject.filter).forEach(key => {
+        if (tableObject.filter[key] === true || tableObject.filter[key] === false) {
+          currentUrl += `;${key}=${tableObject.filter[key]}`;
+        } else {
+          currentUrl += `;${key}=`;
+          tableObject.filter[key].split(',').forEach(item => {
+            currentUrl += `${item},`;
+          });
+          currentUrl = currentUrl.slice(0, -1);
+        }
+      });
+    }
+    currentUrl += ';ms=' + new Date().getTime();
+    window.history.replaceState({}, '', currentUrl);
+  }
+
+  public updateTableObjectWithUrlParams(
+    routeParams: Params,
+    tableObject: TableObject,
+    forceUpdateUrl = false,
+    filterFieldList = []
+  ) {
+    tableObject.pageSize = routeParams.pageSize ? +routeParams.pageSize : DEFAULT_TABLE_PAGE_SIZE;
+    // tslint:disable-next-line: max-line-length
+    tableObject.currentPage = routeParams.currentPage ? +routeParams.currentPage : DEFAULT_TABLE_CURRENT_PAGE;
+    tableObject.sortBy = routeParams.sortBy ? routeParams.sortBy : DEFAULT_TABLE_SORT_BY;
+    tableObject.keywords = routeParams.keywords ? routeParams.keywords : DEFAULT_TABLE_KEYWORDS;
+    filterFieldList.map(field => {
+      if (routeParams[field]) {
+        tableObject.filter[field] = routeParams[field];
+      }
+    });
+
+    // TODO: Get fields from URL.
+
+    if (forceUpdateUrl) {
+      this.updateUrl(tableObject);
+    }
+
+    return tableObject;
+  }
+
+  public navigateUsingParams(navigationObject: TableObject, path: any[]) {
+    const params = this.getNavParamsObj(navigationObject);
+    path.push(params);
+    this.router.navigate(path);
+  }
+
+  public getNavParamsObj(navigationObject: TableObject) {
```

Comment:
General comment for any function.  Make sure to check if params are null, before accessing their values, otherwise we will get exceptions.  It also means we can exit the function immediately which is more efficient (for you optimization guys).

Reply:
The required param is a nav object. Upon creating a nav object, all fields will be filled with a default.